kroxylicious-api/src/test/java/io/kroxylicious/proxy/filter/SafeInvokerTest.java

```diff
@@ -0,0 +1,165 @@
+        FilterContext filterContext = Mockito.mock(FilterContext.class);
+        CompletableFuture<ResponseFilterResult> delegateStage = new CompletableFuture<>();
+        when(invoker.onResponse(any(), anyShort(), any(), any(), any())).thenReturn(delegateStage);
+
+        // when
+        CompletionStage<ResponseFilterResult> stage = safeInvoker.onResponse(API_VERSIONS, (short) 1, header, message,
+                filterContext);
+
+        // then
+        assertThat(stage).isSameAs(delegateStage);
+        verify(invoker).onResponse(API_VERSIONS, (short) 1, header, message, filterContext);
+    }
+
+    @Test
+    void testIfDelegateNotInterestedInResponseSafeInvokerForwards() {
+        // given
+        when(invoker.shouldHandleResponse(API_VERSIONS, (short) 1)).thenReturn(false);
+        ApiVersionsResponseData message = new ApiVersionsResponseData();
+        ResponseHeaderData header = new ResponseHeaderData();
+        FilterContext filterContext = Mockito.mock(FilterContext.class);
+        CompletableFuture<ResponseFilterResult> forwardStage = new CompletableFuture<>();
+        when(filterContext.forwardResponse(header, message)).thenReturn(forwardStage);
+
+        // when
+        CompletionStage<ResponseFilterResult> stage = safeInvoker.onResponse(API_VERSIONS, (short) 1, header, message,
+                filterContext);
+
+        // then
+        assertThat(stage).isSameAs(forwardStage);
+    }
+
+    @Test
+    void testIfDelegateReturnsNullOnRequestThenFailedFutureReturned() {
+        // given
+        when(invoker.shouldHandleRequest(API_VERSIONS, (short) 1)).thenReturn(true);
+        when(invoker.onRequest(any(), anyShort(), any(), any(), any())).thenReturn(null);
+
+        // when
+        CompletionStage<RequestFilterResult> stage = safeInvoker.onRequest(API_VERSIONS, (short) 1, new RequestHeaderData(), new ApiVersionsRequestData(),
+                Mockito.mock(FilterContext.class));
+
+        // then
+        assertThat(stage).isCompletedExceptionally().failsWithin(Duration.ZERO).withThrowableOfType(ExecutionException.class)
+                .withCauseInstanceOf(IllegalStateException.class).havingCause()
+                .withMessageContaining("invoker onRequest returned null for apiKey API_VERSIONS");
+    }
+
+    @Test
+    void testIfDelegateReturnsNullOnResponseThenFailedFutureReturned() {
+        // given
+        when(invoker.shouldHandleResponse(API_VERSIONS, (short) 1)).thenReturn(true);
+        when(invoker.onResponse(any(), anyShort(), any(), any(), any())).thenReturn(null);
+
+        // when
+        CompletionStage<ResponseFilterResult> stage = safeInvoker.onResponse(API_VERSIONS, (short) 1, new ResponseHeaderData(), new ApiVersionsResponseData(),
+                Mockito.mock(FilterContext.class));
+
+        // then
+        assertThat(stage).isCompletedExceptionally().failsWithin(Duration.ZERO).withThrowableOfType(ExecutionException.class)
```

Comment:
`failsWithin` is an interesting assertion. Is there some particular  your trying to demonstrate with that?

Reply:
though I can see I've left a redundant isCompletedExceptionally in there

Reply:
from the javadoc of the deprecation:

> This assertion is deprecated because it relies on hasFailed() semantics which we want to move away from (they are not clear!) and to use failure semantics corresponding to CompletableFuture.get() failing.

and from `hasFailed`:

> Original javadoc
> Verifies that the CompletableFuture has completed exceptionally but has not been cancelled, this assertion is equivalent to:
>  assertThat(future).isCompletedExceptionally()
>                     .isNotCancelled();

Reply:
I think its fine, I just get a bit wary of anything trying to assert timing in unit tests.